### PR TITLE
Introduce bridge relayers pallet

### LINKED ISSUE #1503: On-demand relayer rewards payment
part of #1318 which, I think, may be implemented now, even without final fees decision

As I see #1318, the solution may be:
1) burning tokens when message is being sent && minting tokens directly on (source) parachain;
2) burning tokens when message is being sent && paying bounty from treasury on (source) parachain;
3) moving tokens to treasury when message is being sent && paying bounty from treasury on (source) parachain;
4) some complex schema, involving relaychain.

In any case, we'll need to store total relayer rewards somewhere + a call that will trigger one of those ^^^ options. So I suggest to work on additional call in the messages pallet now and see how it works on RialtoParachain<>Millau.

### PR DESCRIPTION
**Pull request by svyatonik**
closes #1503 

There's still a lot of TODOs here. I'm opening it early so that anyone could ask questions and voice their opinions regarding this new pallet.

Quick intro: we have unsolved #1318 (which I'm planning to discuss once we have some kind of Rococo/Wococo bridge hubs). Till last week I was thinking that we shall not work on that completely, but actually there are parts of the code that may be worked on now. So I've branched off #1503 from #1318. I was going to add new call (`claim_rewards`) and map to the messages pallet (`relayer => reward`) first. But then I've remembered #983, which we also shall solve soon or later. Adding #983 (whatever solution we'll choose) to the existing messages pallet will make it too complex, so imo it makes sense to move all relayer management code (including paying fees and rewards) to the separate pallet, which I'm introducing in this PR.

### CODE REVIEW

**Comment by svyatonik** [comment]
So after this PR, all we need to do to solve  #1318 is (I hope) to properly implement `PaymentProcedure`

File: primitives/relayers/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+# Bridge dependencies
+
+#bp-runtime = { path = "../runtime", default-features = false }
+
+# Substrate Dependencies
+
+frame-support = { git = "https://github.com/paritytech/substrate", branch = "master", default-features = false }
+#frame-system = { git = "https://github.com/paritytech/substrate", branch = "master", default-features = false }
+#sp-core = { git = "https://github.com/paritytech/substrate", branch = "master", default-features = false }
```
Comment (serban300) [inline primitives/relayers/Cargo.toml:19]: Can we delete these commented lines ?

**Review by serban300 — COMMENTED** [review]
Looks very good. I just left a couple of nits.

File: primitives/relayers/Cargo.toml (hunk 1 above)
Comment (svyatonik) [reply primitives/relayers/Cargo.toml:19]: My fault - sorry :(

File: modules/relayers/src/payment_adapter.rs [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+	}
+}
+
+// Update rewards to given relayers, optionally rewarding confirmation relayer.
+fn schedule_relayers_rewards<T: Config>(
+	confirmation_relayer: &T::AccountId,
+	relayers_rewards: RelayersRewards<T::AccountId, T::Reward>,
+	confirmation_fee: T::Reward,
+) {
+	// reward every relayer except `confirmation_relayer`
```
[the remark this answers is not in the document]
Comment (svyatonik) [reply modules/relayers/src/payment_adapter.rs:79]: This comment is related to the loop - we reward confirmation relayer after it is completed

File: modules/relayers/src/payment_adapter.rs [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+				confirmation_relayer,
+				relayers_rewards,
+				GetConfirmationFee::get(),
+			);
+		}
+	}
+}
+
+// Update rewards to given relayers, optionally rewarding confirmation relayer.
+fn schedule_relayers_rewards<T: Config>(
```
[the remark this answers is not in the document]
Comment (svyatonik) [reply modules/relayers/src/payment_adapter.rs:74]: It is up to you :)

**Review by svyatonik — COMMENTED** [review]
Thanks for the review, all your points seems legit to me!

**Comment by serban300** [comment]
> Thanks for the review, all your points seems legit to me!

Ok, thanks ! I'll implement them in a follow-up PR.

File: modules/relayers/src/payment_adapter.rs [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+/// Remember that the reward shall be paid to the relayer.
+fn schedule_relayer_reward<T: Config>(relayer: &T::AccountId, reward: T::Reward) {
+	if reward.is_zero() {
+		return
+	}
+
+	RelayerRewards::<T>::mutate(relayer, |old_reward: &mut Option<T::Reward>| {
+		let new_reward = old_reward.unwrap_or_else(Zero::zero).saturating_add(reward);
+		log::trace!(
+			target: "T::bridge-relayers",
```
[the remark this answers is not in the document]
Comment (svyatonik) [reply modules/relayers/src/payment_adapter.rs:122]: My fault again

**Comment by serban300** [comment]
The comments were addressed in #1547 . Marking this PR as reviewed.

**Pull request merged**